File: go/appencryption/plugins/aws-v1/persistence/memory.go
package persistence

import (
	"context"
	"sort"
	"sync"

	"github.com/godaddy/asherah/go/appencryption"
)

// Verify MemoryMetastore implements the keystore interface.
var _ appencryption.Metastore = (*MemoryMetastore)(nil)

// MemoryMetastore is an in-memory implementation of a Metastore.
// NOTE: It should not be used in production and is for testing only!
type MemoryMetastore struct {
	sync.RWMutex

	Envelopes map[string]map[int64]*appencryption.EnvelopeKeyRecord
}

// NewMemoryMetastore returns a new in-memory metastore.
func NewMemoryMetastore() *MemoryMetastore {
	return &MemoryMetastore{
		Envelopes: make(map[string]map[int64]*appencryption.EnvelopeKeyRecord),
	}
}

// Load retrieves a specific key by id and created timestamp.
// The return value will be nil if not already present.
func (s *MemoryMetastore) Load(_ context.Context, keyID string, created int64) (*appencryption.EnvelopeKeyRecord, error) {
	s.RLock()
	defer s.RUnlock()

	if ret, ok := s.Envelopes[keyID][created]; ok {
		return ret, nil
	}

	return nil, nil
}

// LoadLatest returns the latest key matching the provided ID.
// The return value will be nil if not already present.
func (s *MemoryMetastore) LoadLatest(_ context.Context, keyID string) (*appencryption.EnvelopeKeyRecord, error) {
	s.RLock()
	defer s.RUnlock()

	if keyIDMap, ok := s.Envelopes[keyID]; ok {
		// Sort submap by key since it is the created time
		var createdKeys []int64
		for created := range keyIDMap {
			createdKeys = append(createdKeys, created)
		}

		sort.Slice(createdKeys, func(i, j int) bool { return createdKeys[i] < createdKeys[j] })

		latestCreated := createdKeys[len(createdKeys)-1]
		if ret, ok := keyIDMap[latestCreated]; ok {
			return ret, nil
		}
	}

	return nil, nil
}

// Store attempts to insert the key into the metastore if one is not
// already present. If a key exists, the method will return false. If
// one is not present, the value will be inserted and we return true.
func (s *MemoryMetastore) Store(_ context.Context, keyID string, created int64, envelope *appencryption.EnvelopeKeyRecord) (bool, error) {
	s.Lock()
	defer s.Unlock()

	if _, ok := s.Envelopes[keyID][created]; ok {
		return false, nil
	}

	// If first time, need to initialize nested map
	if _, ok := s.Envelopes[keyID]; !ok {
		s.Envelopes[keyID] = make(map[int64]*appencryption.EnvelopeKeyRecord)
	}

	s.Envelopes[keyID][created] = envelope

	return true, nil
}

File: rust/appencryption/src/persistence/memory.rs
use crate::envelope::EnvelopeKeyRecord;
use crate::error::Result;
use crate::Metastore;

use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

/// An in-memory implementation of the Metastore trait
///
/// This implementation is meant for testing and should not be used in production.
pub struct MemoryMetastore {
    /// Map of key ID -> map of created timestamp -> envelope key record
    envelopes: Arc<RwLock<HashMap<String, HashMap<i64, EnvelopeKeyRecord>>>>,
}

impl MemoryMetastore {
    /// Creates a new MemoryMetastore
    pub fn new() -> Self {
        Self {
            envelopes: Arc::new(RwLock::new(HashMap::new())),
        }
    }
}

impl Default for MemoryMetastore {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl Metastore for MemoryMetastore {
    async fn load(&self, id: &str, created: i64) -> Result<Option<EnvelopeKeyRecord>> {
        let envelopes = self.envelopes.read().unwrap();
        
        if let Some(key_map) = envelopes.get(id) {
            if let Some(record) = key_map.get(&created) {
                return Ok(Some(record.clone()));
            }
        }
        
        Ok(None)
    }
    
    async fn load_latest(&self, id: &str) -> Result<Option<EnvelopeKeyRecord>> {
        let envelopes = self.envelopes.read().unwrap();
        
        if let Some(key_map) = envelopes.get(id) {
            if key_map.is_empty() {
                return Ok(None);
            }
            
            // Find the latest created timestamp
            let latest_created = key_map.keys()
                .max()
                .cloned()
                .unwrap_or(0);
                
            if let Some(record) = key_map.get(&latest_created) {
                return Ok(Some(record.clone()));
            }
        }
        
        Ok(None)
    }
    
    async fn store(&self, id: &str, created: i64, envelope: &EnvelopeKeyRecord) -> Result<bool> {
        let mut envelopes = self.envelopes.write().unwrap();
        
        // Get or create the key map
        let key_map = envelopes
            .entry(id.to_string())
            .or_default();
            
        // Check if the key already exists
        if key_map.contains_key(&created) {
            return Ok(false);
        }
        
        // Store the key
        key_map.insert(created, envelope.clone());
        
        Ok(true)
    }
}